enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,216 @@
+
+The wait period in each validation starts from 5 minutes base time and it is
+incremented by the base wait period after each boot in the verification loop.
+It should be possible to override the base time wait period setting with the
+`MICROSHIFT_WAIT_TIMEOUT_SEC` environment variable in the `/etc/greenboot/greenboot.conf`
+configuration file alongside other [Greenboot Configuration](https://github.com/fedora-iot/greenboot#configuration)
+settings.
+
+### Risks and Mitigations
+
+Edge Devices often have relatively slow networks or occasional loss of network
+connectivity. The risk of declaring a `false positive` on the failed boot may
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, system administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, system administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+system administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+Some 3rd party workloads may become operational before the upgrade is declared
+valid and potentially create or update data on the device. If a rollback is
+performed subsequentially, there is a risk of data loss because the file system
+is reverted to its state before the upgrade. To mitigate this problem, 3rd party
+workloads should never become operational until a boot is declared successful.
+
+```bash
+$ sudo grub2-editenv - list | grep ^boot_success
+boot_success=1
+```
+
+### Drawbacks
+
+The technical implementation of upgrade validation is rather simple as it relies
+on the existing `greenboot` and `rpm-ostree` facilities. However, end users may
+find it annoying to have their Edge Devices rebooted few times until the upgrade
+is either successful or rolled back.
+
+It is possible that certain 3rd party workloads become operational before the
+upgrade is declared successful. This situation may further hurt user experience
+if their Edge Devices are rebooted due to MicroShift core functionality
+validation not passing.
+
+## Design Details
+
+### Open Questions
+
+* Do we need more validations to be implemented on upgrade?
+* Do we need to implement internal validations in MicroShift?
+* Do we need to add special log messages to be parsed by the validation?
+* Is there a need to implement post-boot scripts in case of success or failure? 
```

Comment:
That question seems like something to answer in the upgrade/downgrade enhancement.

Reply:
Should I remove it from here?

Reply:
Yes, I would save it for later.